src/vulkan-renderer/render_graph.cpp

```diff
@@ -493,39 +485,71 @@ void RenderGraph::compile(const RenderResource &target) {
                         image_views.push_back(image->m_image_view);
                     }
 
-                    phys->m_framebuffers.emplace_back(m_device, phys->m_render_pass, image_views, m_swapchain,
-                                                      "Framebuffer");
+                    physical.m_framebuffers.emplace_back(m_device, physical.m_render_pass, image_views, m_swapchain,
+                                                         "Framebuffer");
                 }
             }
         }
     }
 
-    // Allocate and record command buffers.
+    // Allocate command buffers and finished semaphore.
     for (const auto *stage : m_stage_stack) {
-        auto *phys = m_stage_map[stage].get();
-        alloc_command_buffers(stage, phys);
-        record_command_buffers(stage, phys);
+        auto &physical = *stage->m_physical;
+        physical.m_finished_semaphore =
+            std::make_unique<wrapper::Semaphore>(m_device, "Finished semaphore for stage " + stage->m_name);
+        m_log->trace("Allocating command buffers for stage '{}'", stage->m_name);
+        for (std::uint32_t i = 0; i < m_swapchain.image_count(); i++) {
+            physical.m_command_buffers.emplace_back(m_device, m_command_pool,
+                                                    "Command buffer for stage " + stage->m_name);
+        }
     }
 }
 
-void RenderGraph::render(int image_index, VkSemaphore signal_semaphore, VkSemaphore wait_semaphore,
-                         VkQueue graphics_queue) const {
+VkSemaphore RenderGraph::render(std::uint32_t image_index, VkSemaphore wait_semaphore, VkQueue graphics_queue) {
+    // Update dynamic buffers.
+    for (auto &buffer_resource : m_buffer_resources) {
+        if (buffer_resource->m_data_upload_needed) {
+            auto &physical = *buffer_resource->m_physical->as<PhysicalBuffer>();
+            buffer_resource->m_data_upload_needed = false;
```

Comment:
Shouldn't this be moved to after line 521, when the upload succeeded?

Reply:
I would prefer it because it seems more logical.

Reply:
Done